From 75da62874811ac5be77e3e6690c92a00f912aa66 Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <uwu@icenowy.me>
Date: Tue, 25 Mar 2025 14:05:09 +0800
Subject: [PATCH 360/400] drm/verisilicon: dw-hdmi-th1520: fix arbitary pixclk

Currently the TH1520 HDMI PHY glue code only allows certain pixel clock
frequencies, which rejects arbitary resolutions or refresh rates.

Fix this problem by round up to the first known pixel clock that is
higher than the request, like what other PHY glue drivers do.

Signed-off-by: Icenowy Zheng <uwu@icenowy.me>
---
 drivers/gpu/drm/verisilicon/dw_hdmi_tx_phy_gen2.h | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/verisilicon/dw_hdmi_tx_phy_gen2.h b/drivers/gpu/drm/verisilicon/dw_hdmi_tx_phy_gen2.h
index cff284717e7d..d96f16aebf1c 100644
--- a/drivers/gpu/drm/verisilicon/dw_hdmi_tx_phy_gen2.h
+++ b/drivers/gpu/drm/verisilicon/dw_hdmi_tx_phy_gen2.h
@@ -619,10 +619,8 @@ dw_hdmi_tx_phy_gen2_mode_valid(struct dw_hdmi *dw_hdmi, void *data,
 		return MODE_V_ILLEGAL;
 
 	for (i = 0; i < ARRAY_SIZE(mpll_configs); i++) {
-		if (abs(mode->clock - mpll_configs[i].pixelclock) <= 100)
+		if (mode->clock <= mpll_configs[i].pixelclock)
 			return MODE_OK;
-		else if (mode->clock < mpll_configs[i].pixelclock)
-			break;
 	}
 
 	return MODE_NOMODE;
@@ -640,7 +638,7 @@ static int dw_hdmi_tx_phy_gen2_configure(struct dw_hdmi *hdmi, void *data,
 	for (i = 0; i < ARRAY_SIZE(mpll_configs); i++) {
 		config = &mpll_configs[i];
 		/* TODO: add colordepth check later */
-		if (abs(config->pixelclock - pixclk) <= 100)
+		if (pixclk <= config->pixelclock)
 			break;
 	}
 
-- 
2.43.0

